File: launcher/android-launcher/android-launcher/src/main/java/org/asf/connective/tasks/AsyncTaskManager.java
package org.asf.connective.tasks;

import java.util.ArrayList;
import java.util.function.Predicate;
import java.util.stream.Stream;

/**
 * 
 * Internal task thread manager
 * 
 * @author Sky Swimmer
 *
 */
public class AsyncTaskManager {

	private static ArrayList<AsyncTaskThreadHandler> threads = new ArrayList<AsyncTaskThreadHandler>();
	private static ArrayList<AsyncTask> queuedActions = new ArrayList<AsyncTask>();

	static AsyncTask obtainNext() {
		synchronized (queuedActions) {
			if (queuedActions.size() == 0)
				return null;
			return queuedActions.remove(0);
		}
	}

	public static AsyncTask runAsync(Runnable action) {
		AsyncTask tsk = new AsyncTask(action);

		synchronized (threads) {
			// Check if a thread is available, if not, start a new one
			AsyncTaskThreadHandler[] ths = threads.toArray(new AsyncTaskThreadHandler[0]);
			if (!Stream.of(ths).anyMatch(new Predicate<AsyncTaskThreadHandler>() {

				@Override
				public boolean test(AsyncTaskThreadHandler t) {
					return t.isAvailable();
				}

			})) {
				// Start new thread
				AsyncTaskThreadHandler handler = new AsyncTaskThreadHandler();
				threads.add(handler);
				Thread th = new Thread(new Runnable() {
					@Override
					public void run() {
						// Start
						handler.run();

						// End
						synchronized (threads) {
							threads.remove(handler);
						}
					}
				}, "Async task thread");
				th.setDaemon(true);
				th.start();
			}

			// Add task
			synchronized (queuedActions) {
				queuedActions.add(tsk);
			}

			// Return
			return tsk;
		}
	}

}

File: feraltweaks-bootstrap/Modloader/Actions/Pooling/AsyncTaskManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;

namespace FeralTweaks.Actions.Internal.AsyncTasks
{
    /// <summary>
    /// Async Task Manager - Alternate thread pooler
    /// </summary>
    public static class AsyncTaskManager
    {
        private static List<AsyncTaskThread> _threads = new List<AsyncTaskThread>();
        private static List<AsyncTask> _queuedTasks = new List<AsyncTask>();

        internal static AsyncTask ObtainNext()
        {
            lock (_queuedTasks)
            {
                if (_queuedTasks.Count == 0)
                    return null;
                AsyncTask tsk = _queuedTasks[0];
                _queuedTasks.RemoveAt(0);
                return tsk;
            }
        }

        /// <summary>
        /// Runs a action asynchronously
        /// </summary>
        /// <param name="action">Action to run</param>
        /// <returns>AsyncTask</returns>
        public static AsyncTask RunAsync(Action action)
        {
            AsyncTask tsk = new AsyncTask(action);

            // Check if any thread is available
            // If not, start a new one
            lock (_threads)
            {
                AsyncTaskThread[] threads;
                while (true)
                {
                    try
                    {
                        threads = _threads.ToArray();
                        break;
                    }
                    catch { }
                }

                // Check if any thread is available
                if (!threads.Any(t => t.IsAvailable))
                {
                    // Start new thread
                    AsyncTaskThread th = new AsyncTaskThread();
                    _threads.Add(th);
                    Thread thI = new Thread(() =>
                    {
                        // Start
                        th.Run();

                        // End
                        lock (_threads)
                            _threads.Remove(th);
                    });
                    thI.Name = "FeralTweaks Async Task Thread";
                    thI.IsBackground = true;
                    thI.Start();
                }
            }

            // Queue task
            lock (_queuedTasks)
            {
                _queuedTasks.Add(tsk);
            }

            // Return
            return tsk;
        }
    }
}